src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollection.cs

```diff
@@ -57,5 +69,288 @@ private void Initialize(IEdmEntityType entityType)
+                        }
+
+                        PatchItem(structuredObj, origStructuredObj);                        
+                    }
+                    else
+                    {
+                        originalObj.TrySetPropertyValue(propertyName, value);
+                    }
+                }
+            }
+        }
+
+        private IEdmChangedObject HandleFailedOperation(EdmEntityObject changedObj, DataModificationOperationKind operation, EdmStructuredObject originalObj, 
+            List<IEdmStructuralProperty> keys, string errorMessage)
+        {
+            IEdmChangedObject edmChangedObject = null;
+            DataModificationExceptionType dataModificationExceptionType = new DataModificationExceptionType(operation);
+            dataModificationExceptionType.MessageType = new MessageType();
+            dataModificationExceptionType.MessageType.Message = errorMessage;
+
+            // This handles the Data Modification exception. This adds Core.DataModificationException annotation and also copy other instance annotations.
+            //The failed operation will be based on the protocol
+            switch (operation)
+            {
+                case DataModificationOperationKind.Update:
+                    edmChangedObject = changedObj as IEdmChangedObject;
+                    break;
+                case DataModificationOperationKind.Insert:
+                    {
+                        EdmDeltaDeletedEntityObject edmDeletedObject = new EdmDeltaDeletedEntityObject(EntityType);
+                        PatchItem(edmDeletedObject, changedObj);
+
+                        TryGetContentId(changedObj, keys, edmDeletedObject);
+
+                        edmDeletedObject.TransientInstanceAnnotationContainer = changedObj.TransientInstanceAnnotationContainer;
+                        edmDeletedObject.PersistentInstanceAnnotationsContainer = changedObj.PersistentInstanceAnnotationsContainer;
+
+                        edmDeletedObject.TransientInstanceAnnotationContainer.AddResourceAnnotation("Core.DataModificationException", dataModificationExceptionType);
+                        edmChangedObject = edmDeletedObject;
+                        break;
+                    }
+                case DataModificationOperationKind.Delete:
+                case DataModificationOperationKind.Unlink:
+                    {
+                        EdmDeltaEntityObject edmEntityObject = new EdmDeltaEntityObject(EntityType);
+                        PatchItem(originalObj, edmEntityObject);
+
+                        edmEntityObject.TransientInstanceAnnotationContainer = changedObj.TransientInstanceAnnotationContainer;
+                        edmEntityObject.PersistentInstanceAnnotationsContainer = changedObj.PersistentInstanceAnnotationsContainer;
+
+                        edmEntityObject.TransientInstanceAnnotationContainer.AddResourceAnnotation("Core.DataModificationException", dataModificationExceptionType);
```

Comment:
Would be better to introduce an `internal` `const` for `"Core.DataModificationException"` since you're using it in quite a few places

Reply:
+1

---
In reply to: [615060462](https://github.com/OData/WebApi/pull/2455#discussion_r615060462) [](ancestors = 615060462)